Implement network serializer

daleglass opened the pull request:
This is the last part of the build warnings fix. It deals with the issue of glm classes not being trivially copyable, and therefore not being possible to correctly copy with memcpy.

It also makes the code a bit cleaner, by taking a bunch of fiddly offset-tracking code and moving it into a class.

The code was made with the aim to make it as amenable to inlining as possible.

AnimatedMannequin commented:
Here are some builds you may use for testing: 

**macOS-10.15, full**
 - https://public.overte.org/build/overte/PR81/Overte-PR81-6ebed67.dmg

**ubuntu-18.04, full**
 - https://public.overte.org/build/overte/PR81/Overte-PR81-6ebed67.tar.Z
 - https://public.overte.org/build/overte/PR81/Overte-PR81-6ebed67.sh
 - https://public.overte.org/build/overte/PR81/Overte-PR81-6ebed67.tar.gz

**self-hosted_debian-11_aarch64, full**
 - https://public.overte.org/build/overte/PR81/Overte-PR81-6ebed67.sh
 - https://public.overte.org/build/overte/PR81/Overte-PR81-6ebed67.tar.gz
 - https://public.overte.org/build/overte/PR81/Overte-PR81-6ebed67.tar.Z

**windows-2019, full**
 - https://public.overte.org/build/overte/PR81/Overte-PR81-6ebed67.exe

AnimatedMannequin commented:
Here are some builds you may use for testing: 

**ubuntu-18.04, full**
 - https://public.overte.org/build/overte/PR81/Overte-PR81-8b56ee6.tar.gz
 - https://public.overte.org/build/overte/PR81/Overte-PR81-8b56ee6.tar.Z
 - https://public.overte.org/build/overte/PR81/Overte-PR81-8b56ee6.sh

**self-hosted_debian-11_aarch64, full**
 - https://public.overte.org/build/overte/PR81/Overte-PR81-8b56ee6.tar.Z
 - https://public.overte.org/build/overte/PR81/Overte-PR81-8b56ee6.tar.gz
 - https://public.overte.org/build/overte/PR81/Overte-PR81-8b56ee6.sh

**macOS-10.15, full**
 - https://public.overte.org/build/overte/PR81/Overte-PR81-8b56ee6.dmg

**windows-2019, full**
 - https://public.overte.org/build/overte/PR81/Overte-PR81-8b56ee6.exe

AnimatedMannequin commented:
Here are some builds you may use for testing: 

**ubuntu-18.04, full**
 - https://public.overte.org/build/overte/PR81/Overte-PR81-63914b5.sh
 - https://public.overte.org/build/overte/PR81/Overte-PR81-63914b5.tar.Z
 - https://public.overte.org/build/overte/PR81/Overte-PR81-63914b5.tar.gz

**self-hosted_debian-11_aarch64, full**
 - https://public.overte.org/build/overte/PR81/Overte-PR81-63914b5.tar.Z
 - https://public.overte.org/build/overte/PR81/Overte-PR81-63914b5.sh
 - https://public.overte.org/build/overte/PR81/Overte-PR81-63914b5.tar.gz

**macOS-10.15, full**
 - https://public.overte.org/build/overte/PR81/Overte-PR81-63914b5.dmg

**windows-2019, full**
 - https://public.overte.org/build/overte/PR81/Overte-PR81-63914b5.exe

daleglass commented:
Main part is ready for test/review, test still needs updating, which I'll do soon.

AnimatedMannequin commented:
Here are some builds you may use for testing: 

**macOS-10.15, full**
 - https://public.overte.org/build/overte/PR81/Overte-PR81-2f831d0.dmg

**windows-2019, full**
 - https://public.overte.org/build/overte/PR81/Overte-PR81-2f831d0.exe

daleglass commented:
Dug this up, rebased and polished it up. There was an issue with KTX textures I fixed now. At this point, this should work.

This breaks compatibility with the old KTX format, so on first run expect a bunch of warnings about that. That should be fine, it just means the old version cache can't be used, and it'll be fixed by clearing the cache.

JulianGro commented:
> This breaks compatibility with the old KTX format, so on first run expect a bunch of warnings about that. That should be fine, it just means the old version cache can't be used, and it'll be fixed by clearing the cache.

What about pre-baked assets?

daleglass commented:
> What about pre-baked assets?

That's a good point, I'll have to check that out.

daleglass commented:
That was indeed a good point. This should be fixed now.

daleglass commented:
Okay, this is still not quite right. It works, but something in the output still doesn't line up with how the original was.

daleglass commented:
Dev meeting: fixing this could help with GLM issues in #641

HifiExperiments commented on libraries/gpu/src/gpu/Texture_ktx.cpp:
```diff
@@ -27,69 +28,77 @@ using KtxStorage = Texture::KtxStorage;
+ *
+ * @note The usage for textures means breaking compatibility is a bad idea, and that the implementation
+ * should just keep on adding extra data at the bottom of the structure, and remain able to read old
+ * formats. In fact, version 1 KTX can be found in older baked assets.
```
can we add a similar note to this in `deserialize` (and maybe also serialize)?

daleglass replied:
Added.

HifiExperiments commented on libraries/gpu/src/gpu/Texture_ktx.cpp:
```diff
@@ -27,69 +28,77 @@ using KtxStorage = Texture::KtxStorage;
-        memcpy(data, &_samplerDesc, sizeof(Sampler::Desc));
-        data += sizeof(Sampler::Desc);
+        ser << CURRENT_VERSION;
+        ser.addPadding(1);
```
the old `PADDING` was at the end, but these two +1s are in the middle.  shouldn't this be `ser.addPadding(PADDING)` at the end, like for `SphericalHarmonics` below?

HifiExperiments commented on libraries/gpu/src/gpu/Texture_ktx.cpp:
```diff
@@ -27,69 +28,77 @@ using KtxStorage = Texture::KtxStorage;
 
-        return data + PADDING;
+        // The +1 is here because we're adding the CURRENT_VERSION at the top, but since it's declared as a static
+        // const, it's not actually part of the class' size.
```
is this comment still right?  I don't see a +1 in this check

HifiExperiments commented on libraries/gpu/src/gpu/Texture_ktx.cpp:
```diff
@@ -27,69 +28,77 @@ using KtxStorage = Texture::KtxStorage;
-        _usage = Texture::Usage(usageData);
-        data += sizeof(uint32);
+        dsr >> usageData;
+        _usage._flags = gpu::Texture::Usage::Flags(usageData);
```
I would just call the Usage constructor directly in case it does more in the future

HifiExperiments commented on libraries/render-utils/src/Model.cpp:
```diff
@@ -1905,7 +1905,9 @@ void Blender::run() {
 
         // initialize offsets to zero
-        memset(unpackedBlendshapeOffsets.data(), 0, numVertsInMesh * sizeof(BlendshapeOffsetUnpacked));
+        for(BlendshapeOffsetUnpacked &bou : unpackedBlendshapeOffsets) {
```
this has shown up in traces in the past as a particularly expensive area of the code.  how does the performance of this loop compare to the memset?

HifiExperiments replied:
I've address this here: https://github.com/overte-org/overte/pull/1007/files#diff-986f65078e5377507d9b704c012e7935bdcb2e318ed358383ea66656e772d02eR122

by replacing the vec3s with plain floats.  so this change can be reverted

HifiExperiments commented on libraries/shared/src/SerDes.cpp:
```diff
@@ -0,0 +1,85 @@
+
+
+static void dumpHex(QDebug &debug, const char*buf, size_t len) {
+  QString literal;
```
```suggestion
    QString literal;
```

HifiExperiments commented on libraries/shared/src/SerDes.cpp:
```diff
@@ -0,0 +1,85 @@
+//
+//
+//  Created by Dale Glass on 5/6/2022
+//  Copyright 2022 Dale Glass
```
```suggestion
//  Copyright 2024 Overte e.V.
```

HifiExperiments commented on libraries/shared/src/SerDes.h:
```diff
@@ -0,0 +1,953 @@
+//
+//
+//  Created by Dale Glass on 5/6/2022
+//  Copyright 2022 Dale Glass
```
```suggestion
//  Copyright 2024 Overte e.V.
```

HifiExperiments commented on tests/ktx/src/KtxTests.cpp:
```diff
@@ -31,6 +33,19 @@ QString getRootPath() {
     return result;
 }
 
+#if 0
```
remove or implement test?

HifiExperiments reviewed (commented):
nice!  I think maybe the padding being in the middle instead of the end is why old textures are getting messed up?

HifiExperiments commented:
I know everyone is super busy, but since this is blocking the warnings as errors grant work, may I suggest just merging this as is since it works, and we can deal with the unimplemented test at a later time?

ksuprynowicz commented:
Does it still break compatibility with older baked textures, or was that part resolved?

ksuprynowicz commented:
I tried building and testing it, but it uses older protocol, so I can't test it properly.
@daleglass Can you rebase this PR on current master?

ksuprynowicz commented:
The current version hits an assert for me:
```
[12/31 12:50:11] [FATAL] [default] ASSERT failure in Q_UNREACHABLE(): "Q_UNREACHABLE was reached", file /home/ksuprynowicz/overte/overte/libraries/gpu-gl/src/gpu/gl45/GL45BackendTexture.cpp, line 113
```
Backtrace:
```
__pthread_kill_implementation 0x00007ffff049de5c
__pthread_kill_internal 0x00007ffff049debf
__GI_raise 0x00007ffff0449c82
__GI_abort 0x00007ffff04324f0
QMessageLogger::fatal(char const*, ...) const 0x00007ffff60922d1
qt_assert_x(char const*, char const*, char const*, int) 0x00007ffff6091408
gpu::gl45::GL45Backend::syncGPUObject GL45BackendTexture.cpp:113
gpu::gl::GLBackend::setResourceTexture GLBackendPipeline.cpp:304
gpu::gl::GLBackend::bindResourceTexture GLBackendPipeline.cpp:267
gpu::gl::GLBackend::do_setResourceTextureTable GLBackendPipeline.cpp:329
gpu::gl::GLBackend::renderPassDraw GLBackend.cpp:497
gpu::gl::GLBackend::render GLBackend.cpp:569
gpu::Context::executeFrame Context.cpp:146
OpenGLDisplayPlugin::present OpenGLDisplayPlugin.cpp:716
PresentThread::run OpenGLDisplayPlugin.cpp:198
<unknown> 0x00007ffff60de9e1
start_thread 0x00007ffff049c112
__GI___clone3 0x00007ffff051a8f8
```

HifiExperiments commented:
@ksuprynowicz ah apologies, there was still one bug leftover.  I've rebased this and fixed that here: #1293 

and it does maintain compatibility with old baked textures!
